Use keyboard-types crate

pyfisch opened the pull request:
Have embedders send DOM keys to servo and use a strongly typed KeyboardEvent
from the W3C UI Events spec. All keyboard handling now uses the new types.

Introduce a ShortcutMatcher to recognize key bindings. Shortcuts are now
recognized in a uniform way.

Updated the winit port.
Updated webdriver integration.

part of #20331 

What this PR does:

* allow the use non-ASCII keyboards for text input
* decouple keyboard event "key" from "code" (key meaning vs location)

What this PR does not do:

* completely improve keyboard events send from winit and webdriver
* add support for CompositionEvent or IME

Notes:

* The winit embedder does not send keyup events for printable keys (this is a regression)
* keyboard-types is on crates.io because I believe it to be useful outside of servo. If you prefer I can add a copy in this repo.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/21881)
<!-- Reviewable:end -->

pyfisch committed:
Use keyboard-types crate

Have embedders send DOM keys to servo and use a strongly typed KeyboardEvent
from the W3C UI Events spec. All keyboard handling now uses the new types.

Introduce a ShortcutMatcher to recognize key bindings. Shortcuts are now
recognized in a uniform way.

Updated the winit port.
Updated webdriver integration.

part of #20331

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/webdriver_server/keys.rs, components/constellation/lib.rs, components/script/textinput.rs, components/script/dom/document.rs, components/script/lib.rs and 7 more
 * @cbrewster: components/constellation/lib.rs, components/constellation/Cargo.toml, components/constellation/constellation.rs
 * @jgraham: components/webdriver_server/keys.rs, components/webdriver_server/lib.rs, components/webdriver_server/Cargo.toml
 * @paulrouget: ports/servo/glutin_app/keyutils.rs, components/compositing/lib.rs, components/constellation/lib.rs, components/servo/lib.rs, ports/servo/non_android_main.rs and 7 more
 * @kichjang: components/script/textinput.rs, components/script/dom/document.rs, components/script_traits/lib.rs, components/script/lib.rs, components/script/Cargo.toml and 3 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

pyfisch commented:
@bors-servo test=wpt

pyfisch commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit a2b9bedf204cdfb9e80d2bd48aa9f54449feaaf5 with merge 556739465fae934b097bf28aff4a436257f78991...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10042)

pyfisch commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 23a84a661950c7ac8816d52b5a42ed72782c56d5 with merge fd0072034a13d2366d02939a60970615a6a83052...

bors-servo commented:
:boom: Test timed out

pyfisch commented:
Questions for reviewers:

* keyboard-types
  * naming: `KeyState` or just `State`?
  * should I remove the `CompositionEvent`? It is not used yet but will be used for composing keys (and IMEs?).
  * `ShortcutMatcher` currently takes all modifier keys into account. But some keys like <kbd>NumLock</kbd> are not usually part of shortcuts and they are locking keys so they will prevent any other shortcuts from working. Which keys are usually used in shortcuts? (<kbd>Shift</kbd>, <kbd>Control</kbd>, <kbd>Alt</kbd>, <kbd>Meta</kbd> only?)
  * naming: is `Key::Character` a good name as it contains a `String` and not a `char`?
* webdriver
  * What is this mysterious Return key (`\u{E006}`) referring to? UI Events key Values only has Enter. (see also w3c/webdriver#1271)
* Do all keyboard shortcuts work? (Especially on Mac)
  * <kbd>Control</kbd>+<kbd>Shift</kbd>+<kbd>3</kbd> (capture webrender does not work because the characters emitted by winit for <kbd>Control</kbd>+* are messed up
  
(to be continued ;-) )

When someone reviews this (unfortunately large) PR Id suggest to start with the keyboard-types crate, then continue to the DOM KeyboardEvent and textinput file and last review the embedders winit and webdriver.

CYBAI commented on components/script/dom/document.rs:
```diff
@@ -1348,68 +1349,61 @@ impl Document {
-        modifiers: KeyModifiers,
-    ) {
+    pub fn dispatch_key_event(&self,
+                              keyboard_event: ::keyboard_types::KeyboardEvent) {
```
nit: indentation

```rust
pub fn dispatch_key_event(
    &self,
    keyboard_event: ::keyboard_types::KeyboardEvent
) {
```

CYBAI commented on components/script/dom/document.rs:
```diff
@@ -1438,16 +1427,16 @@ impl Document {
-            match key {
-                Key::Space if state == KeyState::Released => {
+            match keyboard_event.key {
+                Key::Character(ref letter) if letter == " " && keyboard_event.state == KeyState::Up => {
```
Due to changing from `Key::Space` to `Key::Character` with checking `" "`, I'm just curious, if the space is in `full width`, would it be handled in this matching arm? 🤔

CYBAI replied:
Ah, I see. Sorry for my misunderstanding and, yes, it will match for ASCII only so it should be fine to match the `" "` here like original way.

https://github.com/servo/servo/blob/511dfe8e1b062324fbfde0aac8d16cc5e20cf8ce/components/webdriver_server/keys.rs#L13-L15

CYBAI commented on components/script/script_thread.rs:
```diff
@@ -2834,13 +2834,13 @@ impl ScriptThread {
-                document.dispatch_key_event(ch, key, state, modifiers);
-            },
+                document.dispatch_key_event(key_event);
+            }
```
nit: I think `rustfmt` would ask to add the trailling comma back

pyfisch replied:
Thanks!

pyfisch commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit e8d944762898769c53fbbfa591c2353cda76065a with merge d566217dcef6a351615d05395ddba9250cce65f2...

bors-servo commented:
:broken_heart: Test failed - [android-x86](http://build.servo.org/builders/android-x86/builds/839)

pyfisch commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 0ccaa7e1a9e9bf9472d869576019b9cda350ad87 with merge e5778c6c6d13f5ff629dded3fe11899fa8c88a0f...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/12726), [android-x86](http://build.servo.org/builders/android-x86/builds/840), [arm32](http://build.servo.org/builders/arm32/builds/12909), [arm64](http://build.servo.org/builders/arm64/builds/12948), [linux-dev](http://build.servo.org/builders/linux-dev/builds/13020), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10049), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/10071), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/13042), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5500), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5504), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9763), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9813), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5569), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5506), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/Q-eI7ASjQKCoM5X43mZukA), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8404)
State: approved= try=True

paulrouget commented:
Wow, this is amazing work. /cc @fabricedesre

paulrouget commented:
> When someone reviews this (unfortunately large) PR Id suggest to start with the keyboard-types crate, then continue to the DOM KeyboardEvent and textinput file and last review the embedders winit and webdriver.

I'd be happy to look at this. Assigning myself.

paulrouget commented:
> BTW selecting text with the arrow keys and shift can cause an assertion failure (select some chars to the left and then press shift+right)

Is that a regression?

pyfisch commented:
> > BTW selecting text with the arrow keys and shift can cause an assertion failure (select some chars to the left and then press shift+right)
> 
> Is that a regression?

No, at least not one introduced by this commit. It is also present in master. Filed #21891. (Probably "Good first PR" or "E-Easy")

paulrouget commented:
About keyboard-types:

- can you document how to use convert.py in the README (and add the url of the html files)
- maybe do not include key.rs and code.rs in the repo and generate them at compilation time?
- would it be possible to generation location.rs and modifiers.rs as well?
- why serializable?

> Introduce a ShortcutMatcher to recognize key bindings. Shortcuts are now recognized in a uniform way

- what do you mean in a uniform way?
- can you explain the benefit of ShortcutMatcher and its chaining system (`shortcut().shortcut().otherwise()`) vs. matching the events
- can you explain the goal of the shortcut's value?

>   * naming: `KeyState` or just `State`?

KeyState works.

>   * should I remove the `CompositionEvent`? It is not used yet but will be used for composing keys (and IMEs?).

I was thinking that too. Maybe keep it. It's actually useful to see how it will work in the future. Maybe file an issue to describe what it would require to make that work and add a comment linking to that issue.

>   * `ShortcutMatcher` currently takes all modifier keys into account. But some keys like NumLock are not usually part of shortcuts and they are locking keys so they will prevent any other shortcuts from working. Which keys are usually used in shortcuts? (Shift, Control, Alt, Meta only?)

Yes. Ignore `NumLock`.

>   * naming: is `Key::Character` a good name as it contains a `String` and not a `char`?

That's fine.

paulrouget commented on ports/servo/glutin_app/window.rs:
```diff
@@ -408,56 +408,68 @@ impl Window {
+            modifiers: self.key_modifiers.get(),
+            repeat: false,
+            is_composing: false,
+        };
```
Maybe introduce `keyutils::keyevent_from_winit() -> KeyboardEvent` instead of having all these get_servo__*_from_winit.

paulrouget commented:
Maybe add a feature to keyboard-types that would add a way to translate winit event to Keyboard events. Ideally, we would be able to pass the `ReceivedCharacter` and `KeyboardInput` events to keyboard-types and get relevant translated events?

What do you think?

paulrouget commented on components/webdriver_server/keys.rs:
```diff
@@ -2,184 +2,105 @@
-        _ => None,
+// spec: https://w3c.github.io/webdriver/#keyboard-actions
+// normalised (sic) as in british spelling
+fn get_normalised_key_value(key: char) -> Key {
```
Could that be an utility function part of keyboard-types?

pyfisch replied:
This is a webdriver specific function. If you want to include webdriver specific functions in keyboard-types I'd directly put `keycodes_to_keys` in keyboard-types which converts a string of input characters into a sequence of keydown and keyup events.

paulrouget commented on components/script/dom/keyboardevent.rs:
```diff
@@ -118,690 +103,94 @@ impl KeyboardEvent {
 
 // https://w3c.github.io/uievents/#legacy-key-models
-fn key_keycode(key: Key) -> u32 {
+pub fn key_keycode(key: &Key) -> u32 {
```
Could we move this to keyboard-types?

pyfisch replied:
sure.

pyfisch commented:
> * can you document how to use convert.py in the README (and add the url of the html files)

check

> * maybe do not include key.rs and code.rs in the repo and generate them at compilation time?

The convert.py file is a quick and dirty Python script. I don't think It is a good idea to require Python 3, the requests lib and an internet connection just to build keyboard-types. (I don't want to RIIR.)

> * would it be possible to generation location.rs and modifiers.rs as well?

I think it's pointless to generate location.rs as it will likely never be changed. The modifiers.rs file could change but there weren't enough variants to warrant a generator.



>> Introduce a ShortcutMatcher to recognize key bindings. Shortcuts are now recognized in a uniform way

> * what do you mean in a uniform way?

Currently there are at least three different ways used in servo to match a shortcut like <kbd>Control</kbd> + <kbd>C</kbd>:

1. `(KeyModifiers::CONTROL, None, Key::C, _) => ...`
2. `(mods, None, Key::C, _) if is_control_key(mods) => ...`
3. `(mods, None, Key::C, _) if mods.contains(KeyModifiers::CONTROL) => ...`

While the intention is the same the exact behavior differs. The ShortcutMatcher offers only one way to match a given keyboard shortcut.

> * can you explain the benefit of ShortcutMatcher and its chaining system (`shortcut().shortcut().otherwise()`) vs. matching the events

Matching gets really awkward once you are no longer matching simple values.
As Key may contain a Character string I can no longer say `match key { Key::Character("a") => ...` 
but need to write `match key { Key::Character(ref text) if text == "a" || text == "A"`.
Then you need to add modifier keys. In a match arm you can't write `Modifiers::CONTROL | Modifiers::Alt`
for <kbd>Control</kbd>+<kbd>Alt</kbd> but need to introduce a new constant to match
`const CONTROL_AND_ALT = Modifiers::CONTROL | Modifiers::Alt`.

Now one can argue that a function `test_shortcut(Modifiers::CONTROL, 'c', event)`  should be defined
that contains all the logic ShortcutMatcher has. This would be called in a long if-else-if-else-else chain.
Here I made the largely stylistic choice to use a chaining system.

(Note that the if-else chain would still need a nested if to test if the key was pressed or released in which
case the event is ignored)

> * can you explain the goal of the shortcut's value?

The goal is to have a return value like with a match statement.

> * can you explain the benefit of ShortcutMatcher and its chaining system (`shortcut().shortcut().otherwise()`) vs. matching the events
> 
> * can you explain the goal of the shortcut's value? (edit: just saw how it's used in textinput.rs)



> > * `ShortcutMatcher` currently takes all modifier keys into account. But some keys like NumLock are not usually part of shortcuts and they are locking keys so they will prevent any other shortcuts from working. Which keys are usually used in shortcuts? (Shift, Control, Alt, Meta only?)
> 
> Yes. Ignore `NumLock`.

I decided to ignore all modifier keys except Shift, Control, Alt, Meta for now.


> Maybe add a feature to keyboard-types that would add a way to translate winit event to Keyboard events. Ideally, we would be able to pass the `ReceivedCharacter` and `KeyboardInput` events to keyboard-types and get relevant translated events?
> 
> What do you think?

I want keyboard-types to be a crate to define types and utities like shortcut matching. This means the crate has few dependencies and can be used in different contexts. Adding winit integration would go against this. More importantly I don't think winit keyboard events are sufficient for a browser. If I recall correctly the keyboard events were copied from SDL2 a game engine with very different needs than a web browser. Additionally there are many problems with winits keyboard mapping (like Control+*) or non-ASCII input as shown by the numerous winit issues. I'd prefer to write proper keyboard handling for servo on different OS over using winit events forever. Maybe xkbcommon can be used on Linux.

pyfisch commented:
@paulrouget I updated keyboard-types with your review comments: https://github.com/pyfisch/keyboard-types/commit/db249f029e72f9d6c31c8d5f8cdc637b708dc05c

pyfisch committed:
Minor keyboard updates

Support combined character input from winit.
Make use of utility methods from keyboard-types.
Remove prinatable attribute of KeyboardEvent.

pyfisch commented:
Update PR. Addressed the review comments.

One more question: While the [function key table](https://w3c.github.io/uievents-key/#keys-function) only enumerates F1-F12 and Soft1-Soft4 it states that higher numbers such as F24 are equally valid. Should the keyboard types crate support those keys? I am not aware of any modern keyboard with such keys. Instead of variants for F1 to F12 I'd use a single variant `F(u8)` to represent all possible function keys.

pyfisch commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit c8619424f20f5ad58b5e36624464605a3111de66 with merge a8ce3527003929e74eea6bdd80feb0d2e9a09c3a...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/10193)

pyfisch commented:
@bors-servo try=wpt

jdm commented:
@bors-servo retry

bors-servo commented:
:boom: Test timed out

jdm commented:
It was green.

paulrouget commented:
I'm happy with both this PR and keyboard-types. This is a significant improvement. Thank you.
I haven't had a chance to test on MacOS though because of https://github.com/servo/mozjs/issues/156.

@bors-servo r+

bors-servo commented:
:pushpin: Commit c861942 has been approved by `paulrouget`

<!-- @bors-servo r=paulrouget c8619424f20f5ad58b5e36624464605a3111de66 -->

bors-servo commented:
:hourglass: Testing commit c8619424f20f5ad58b5e36624464605a3111de66 with merge 9a0404ac5fc1c463aa68bb0c68d709a61157c2bf...

nox commented:
I'm confused, why are we using a prerelease version of this crate?

nox commented:
I'm confused, why are we using a prerelease version of this crate?

nox commented:
I'm confused, why are we using a prerelease version of this crate?

pyfisch commented:
> I'm confused, why are we using a prerelease version of this crate?

Because I wanted some feedback on the keyboard-types crate without publishing a new release version for keyboard-types.
